zipkintracing/tracing.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return err
+	}
+	t.client = client
+	t.tracer = tracer
+	return nil
+}
+
+// NewTracer returns instance of tracer with client initialized
+func NewTracer(zipkinURL, serviceName string) (Tracer, error) {
+	tracer := &tracer{
+		ReportingURL: zipkinURL,
+		ServiceName:  serviceName,
+	}
+	err := tracer.init()
+	if err != nil {
+		return nil, err
+	}
+	return tracer, nil
+}
+
+// DoHTTP is a http zipkin tracer implementation of HTTPDoer
+func (t *tracer) DoHTTP(c echo.Context, r *http.Request) (*http.Response, error) {
+	req := r.WithContext(c.Request().Context())
+	return t.client.DoWithAppSpan(req, r.Method)
+}
+
+// TraceFunc wraps function call with span so that we can trace time taken by func, eventContext only provided if we want to store trace headers
+func (t *tracer) TraceFunc(c echo.Context, spanName string, tags map[string]string) func() {
+	span, _ := t.tracer.StartSpanFromContext(c.Request().Context(), spanName)
+	for key, value := range tags {
+		span.Tag(key, value)
+	}
+	finishSpan := func() {
+		span.Finish()
+	}
+
+	return finishSpan
+}
+
+func (t *tracer) GetTraceProxyFunc(tags map[string]string) echo.MiddlewareFunc {
+	return t.GetTraceProxyFuncWithConfig(tags, DefaultTraceProxyFuncConfig)
+}
+
+// GetTraceProxyFunc takes desire tags and returns middleware used to trace proxy calls
+func (t *tracer) GetTraceProxyFuncWithConfig(tags map[string]string, config TraceProxyFuncConfig) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			var parentContext model.SpanContext
+			if span := zipkin.SpanFromContext(c.Request().Context()); span != nil {
+				parentContext = span.Context()
+			}
+			span := t.tracer.StartSpan(fmt.Sprintf("C %s %s", c.Request().Method, "reverse proxy"), zipkin.Parent(parentContext))
+			for key, value := range tags {
+				span.Tag(key, value)
+			}
+			defer span.Finish()
+			ctx := zipkin.NewContext(c.Request().Context(), span)
+			c.Request().WithContext(ctx)
+			b3.InjectHTTP(c.Request())(span.Context())
+			nrw := NewResponseWriter(c.Response().Writer)
+			next(c)
+
+			if nrw.Size() > 0 {
+				zipkin.TagHTTPResponseSize.Set(span, strconv.FormatInt(int64(nrw.Size()), 10))
+			}
+			if nrw.Status() < 200 || nrw.Status() > 299 {
+				statusCode := strconv.FormatInt(int64(nrw.Status()), 10)
+				zipkin.TagHTTPStatusCode.Set(span, statusCode)
+				if nrw.Status() > 399 {
+					zipkin.TagError.Set(span, statusCode)
+				}
+			}
+			return nil
+		}
+	}
+}
+
+func (t *tracer) GetTraceServerFunc(tags map[string]string) echo.MiddlewareFunc {
+	return t.GetTraceServerFuncWithConfig(tags, DefaultTraceServerFuncConfig)
+}
+
+// GetTraceServerFunc takes desire tags and returns middleware used to trace server call
+func (t *tracer) GetTraceServerFuncWithConfig(tags map[string]string, config TraceServerFuncConfig) echo.MiddlewareFunc {
```

Comment:
Should it be method on `tracer` struct?  

Maybe having a function to take tracer as argument would allow developer more control over how tracer instance is setup/configured. 
```go
func GetTraceServerFuncWithConfig(
   tracer *zipkin.Tracer, 
   tags map[string]string, 
   config TraceServerFuncConfig
) echo.MiddlewareFunc {
```

because Tracer has (more) options if you look its creation method `func NewTracer(rep reporter.Reporter, opts ...TracerOption)`

or add tracer into `TraceServerFuncConfig` as it usually is done.

```go
	TraceServerFuncConfig struct {
		Skipper middleware.Skipper
		Tracer  *zipkin.Tracer
	}
```

Reply:
I agree, the tracer struct I was creating was too opinionated and didn't allow any modifications. I took most of your suggestions and updated accordingly.